src/common/base/comparable.ts

```diff
@@ -8,33 +8,49 @@
  * http://polymer.github.io/PATENTS.txt
  */
 
-import {assert} from '../../platform/assert-web.js';
-import {Dictionary} from '../hot.js';
 
-export function compareNulls<T>(o1: T | null, o2: T | null) {
+import {Dictionary} from '../../runtime/hot.js';
+import {checkDefined} from './preconditions.js';
+
+/**
+ * @fileoverview Utility code to compare objects and primitives.
+ */
+
+// TODO(lindner): this needs some work to make it more like the java
+// comparable interface or python __lt__
+
+export interface Comparable<T> {
+  _compareTo<T>(other: Comparable<T>): number;
+}
+
+
+// TODO(lindner): move the utility code outside, break dep on Dictionary
+
+export function compareNulls<T>(o1: T | null, o2: T | null): number {
   if (o1 === o2) return 0;
   if (o1 === null) return -1;
   return 1;
 }
 
-export function compareStrings(s1: string | null, s2: string | null) {
+export function compareStrings(s1: string | null, s2: string | null): number {
   if (s1 == null || s2 == null) return compareNulls<string>(s1, s2);
   return s1.localeCompare(s2);
 }
 
-export function compareNumbers(n1: number, n2: number) {
-  if (n1 == null || n2 == null) return compareNulls<number>(n1, n2);
+export function compareNumbers(n1: number | null, n2: number | null): number {
+  if (n1 === null || n2 === null) return compareNulls<number>(n1, n2);
```

Comment:
We should be careful about strong vs weak equality. 
Can the arguments every be `undefined`? If so, then this will break this logic. 
There are two options: 
- handle `undefined` seperately, have these functions accept `undefined` as a possible argument.
- use weak equality. 


```
(base) alxrsngrtn-macbookpro:emsdk alxrsngrtn$ node
> null === undefined
false
> undefined === null
false
> null == undefined
true
> undefined == null
true
```

Reply:
Also, my quick test could be wrong in typescript-land.

Reply:
at some point It might make sense to move null handling out of these methods entirely and only allow for explicit numerical compare.

Guava's [Ordering](https://guava.dev/releases/21.0/api/docs/com/google/common/collect/Ordering.html) comes to mind.

For now undefined values are not explicitly handled.  This is not needed or used by current use cases:

- Comparing Array lengths in Dictionary.compare 
-comparing version numbers in StorageProviderBase.